Tasks/Maven/maven2.js

```diff
@@ -48,40 +51,149 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
-      var buildFolder = tl.getVariable('agent.buildDirectory');
-      var allFiles = tl.find(buildFolder);
-      var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, { matchBase: true });
+    if (publishJUnitResults == 'true') {
+        //check for pattern in testResultsFiles
+        if (testResultsFiles.indexOf('*') >= 0 || testResultsFiles.indexOf('?') >= 0) {
+            tl.debug('Pattern found in testResultsFiles parameter');
+            var buildFolder = tl.getVariable('agent.buildDirectory');
+            var allFiles = tl.find(buildFolder);
+            var matchingTestResultsFiles = tl.match(allFiles, testResultsFiles, {
+                matchBase: true
+            });
+        } else {
+            tl.debug('No pattern found in testResultsFiles parameter');
+            var matchingTestResultsFiles = [testResultsFiles];
+        }
+
+        if (!matchingTestResultsFiles) {
+            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
+            return 0;
+        }
+
+        var tp = new tl.TestPublisher("JUnit");
+        tp.publish(matchingTestResultsFiles, false, "", "");
     }
-    else {
-      tl.debug('No pattern found in testResultsFiles parameter');
-      var matchingTestResultsFiles = [testResultsFiles];
+}
+
+function getSonarQubeRunner() {
+    var sqAnalysisEnabled = tl.getInput('sqAnalysisEnabled', true);
+
+    if (sqAnalysisEnabled != 'true') {
+        console.log("SonarQube analysis is not enabled.");
+        return;
     }
 
-    if(!matchingTestResultsFiles) {
-      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');  
-      return 0;
+    console.log("SonarQube analysis is enabled");
+    var sqArguments;
+    var sqEndpoint = getEndpointDetails("sqConnectedServiceName");
+    var sqDbDetailsRequired = tl.getInput('sqDbDetailsRequired', true);
+
+    if (sqDbDetailsRequired == 'true') {
+        var sqDbUrl = tl.getInput('sqDbUrl', false);
+        var sqDbUsername = tl.getInput('sqDbUsername', false);
+        var sqDbPassword = tl.getInput('sqDbPassword', false);
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password, sqDbUrl, sqDbUsername, sqDbPassword);
+    } else {
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password);
     }
 
-    var tp = new tl.TestPublisher("JUnit");
-    tp.publish(matchingTestResultsFiles, false, "", "");
-  } 
+    sqArguments = mavenOptions + " " + sqArguments;
+    tl.debug("Running Maven with goal sonar:sonar and options " + sqArguments);
+
+    var mvnsq = new tl.ToolRunner(mvntool);
+    mvnsq.arg('-f');
+    mvnsq.arg(mavenPOMFile);
+    mvnsq.arg(sqArguments);
+    mvnsq.arg("sonar:sonar");
+
+    return mvnsq;
 }
 
+function getEndpointDetails(inputFieldName) {
```

Comment:
We're using a generic endpoint to specify the SonarQube host, username & password